network/src/application/interface.rs

```diff
@@ -25,36 +30,42 @@ pub trait NetworkInterface {
     fn sender(&self) -> Self::Sender;
 
     /// Retrieve only connected peers
-    fn connected_peers(&self) -> HashMap<PeerId, PeerInfo> {
-        self.filtered_peers(|(_, peer_info)| peer_info.status == PeerState::Connected)
+    fn connected_peers(&self, network_id: NetworkId) -> HashMap<PeerNetworkId, PeerInfo> {
+        self.filtered_peers(network_id, |(_, peer_info)| {
+            peer_info.status == PeerState::Connected
+        })
     }
 
     /// Filter peers with according `filter`
     fn filtered_peers<F: FnMut(&(&PeerId, &PeerInfo)) -> bool>(
         &self,
+        network_id: NetworkId,
         filter: F,
-    ) -> HashMap<PeerId, PeerInfo> {
-        self.peer_metadata_storage().read_filtered(filter)
+    ) -> HashMap<PeerNetworkId, PeerInfo> {
+        self.peer_metadata_storage()
+            .read_filtered(network_id, filter)
     }
 
     /// Retrieve PeerInfo for the node
-    fn peers(&self) -> HashMap<PeerId, PeerInfo> {
-        self.peer_metadata_storage().read_all()
+    fn peers(&self, network_id: NetworkId) -> HashMap<PeerNetworkId, PeerInfo> {
+        self.peer_metadata_storage().read_all(network_id)
     }
 
     /// Insert application specific data
-    fn insert_app_data(&self, peer_id: PeerId, data: Self::AppData);
+    fn insert_app_data(&self, app_data_key: Self::AppDataKey, data: Self::AppData);
 
     /// Removes application specific data
-    fn remove_app_data(&self, peer_id: &PeerId);
+    fn remove_app_data(&self, app_data_key: &Self::AppDataKey);
 
     /// Read application specific data
-    fn read_app_data(&self, peer_id: &PeerId) -> Option<Self::AppData>;
+    fn read_app_data(&self, app_data_key: &Self::AppDataKey) -> Option<Self::AppData>;
 
     /// Write application specific data, allows for read before write operations
-    fn write_app_data<F: FnOnce(&mut Entry<PeerId, Self::AppData>) -> Result<(), PeerError>>(
+    fn write_app_data<
+        F: FnOnce(&mut Entry<Self::AppDataKey, Self::AppData>) -> Result<(), PeerError>,
+    >(
```

Comment:
one question i have is: do applications have to manage garbage collecting the app data?

at least thinking about state sync and mempool, I think you would want app_data to reflect discovery 1:1. in other words, I have an app data for each discovered peer. even if the peer disconnects, I still have the data. but if the peer is removed from discovery (e.g. leaves the validator set), I would want their app_data purged as well.

this would be in contrast to syncing it with _connected_ peers. you wouldn't want a bad guy to just reconnect and clear his bad state sync reputation. maybe we do discovered U connected?

Reply:
At this moment, it's simplistic based on disconnections (and handled per application).  I agree with you it would be better in the future to keep the `discovered peers` around permanently, and only discard `incoming peers` e.g. non-known clients.

It's a first step right now, as discovered peers aren't yet threaded through the peer metadata state.